### ISSUE TITLE
[Bug]: Inconsistent results from `LinearMCObjective`

### ISSUE THREAD

**Issue by AdrianSosic** [pytorch/botorch#2681]
### What happened?

I'm currently playing around with mechanisms for function minimization as suggested [here](https://github.com/pytorch/botorch/issues/2316#issuecomment-2085964607) and noticed some strange behavior. Potentially I'm overlooking some crucial aspect or maybe there is simply a flaw in my approach. But what I observe is that the acquisition values obtained when using a `LinearMCObjective` do not match up when maximizing a GP or minimizing an equivalent GP trained on the negated target values. 

### Please provide a minimal, reproducible example of the unexpected behavior.

Here minimal example inspired by the code from the landing page. With `qProbabilityOfImprovement`, the differences are even more striking.

```python
import random

import numpy as np
import torch
from botorch.acquisition import qLogExpectedImprovement
from botorch.acquisition.objective import LinearMCObjective
from botorch.fit import fit_gpytorch_mll
from botorch.models import SingleTaskGP
from botorch.models.transforms import Normalize, Standardize
from gpytorch.mlls import ExactMarginalLogLikelihood

train_X = torch.rand(5, 2, dtype=torch.double) * 2
Y = 1 - torch.linalg.norm(train_X - 0.5, dim=-1, keepdim=True)
Y = Y + 0.1 * torch.randn_like(Y)


def fix_seed(seed=0):
    torch.manual_seed(seed)
    random.seed(seed)
    np.random.seed(seed)


def run(minimize: bool):
    if minimize:
        train_Y = -Y
        obj = LinearMCObjective(weights=torch.tensor([-1.0]))
        best_f = -Y.min()
    else:
        train_Y = Y
        obj = LinearMCObjective(weights=torch.tensor([1.0]))
        best_f = Y.max()

    gp = SingleTaskGP(
        train_X=train_X,
        train_Y=train_Y,
        input_transform=Normalize(d=2),
        outcome_transform=Standardize(m=1),
    )
    mll = ExactMarginalLogLikelihood(gp.likelihood, gp)

    fix_seed(0)
    fit_gpytorch_mll(mll)
    posterior = gp(train_X)

    acqf = qLogExpectedImprovement(model=gp, best_f=best_f, objective=obj)

    fix_seed(0)
    acqf_value = acqf(train_X.unsqueeze(1))
    return posterior, acqf_value


posterior_max, acqf_max = run(False)
posterior_min, acqf_min = run(True)

# The predictions are equivalent
assert torch.equal(posterior_max.mean, -posterior_min.mean)
assert torch.equal(posterior_max.stddev, posterior_min.stddev)

# But the acquisition values do not match
print(acqf_max.tolist())
print(acqf_min.tolist())
```

Sometimes, the number "sort of match up" like:
```python
[-41.65546159698638, -44.19458355943439, -5.794877518131068, -3.9572762719339836, -43.939276772549725]
[-41.54524809215892, -44.165510927608004, -5.034962168263423, -3.5053637720690194, -43.90618926555327]
```

But oftentimes, there are very significant differences:
```python
[-41.253846088273036, -42.4522252849054, -43.566397044168006, -4.586150277498735, -43.92969884779241]
[-5.776667325514115, -41.053594326714006, -42.89699038185324, -1.3871302702025305, -43.38882095231844]
```

Is something wrong with my logic?

### Please paste any relevant traceback/logs produced by the example provided.

```shell

```

### BoTorch Version

0.12.0

### Python Version

3.10.14

### Operating System

macOS

### Code of Conduct

- [x] I agree to follow BoTorch's Code of Conduct

**Comment by saitcakmak** [comment]
Hi @AdrianSosic. The issue here is `best_f = -Y.min()`. `best_f` is compared to objective values, so it should be `obj(train_Y).max() = Y.max()` even in the `minimize` case.

**Comment by AdrianSosic** [comment]
Hi @saitcakmak, thanks for the quick response. I do have my trouble understanding the logic behind, though. These are my concerns:
1. First a very objective observation: If I use `Y.max()` also in the `minimize` case, the numbers are
    * generally "closer" than in the original code and in most cases there is no big difference
    * but if you run it several times, you'll still see a drastic gap somtimes. Here the outcome of one such run (see second entry):
      ```python
      [-44.31642632073237, -39.24304065889777, -43.412800363898974, -42.622768652200996, -4.257021584505255]
      [-44.31624660730143, -10.235137206528035, -43.412491991884835, -42.62219805342741, -4.2532742249170665]
      ```
2. Even though the numbers are closer, I expect that there should be an _exact_ match when all random seeds are properly controlled. Assuming `Y.max()` is the correct thing to do, what would explain the numerical differences?
3. Just conceptually, I don't understand how `Y.max()` could possibly be the right choice 🤔 If we talk about function minimization using expected improvement, then the minimum of my data is what literally defines the baseline to beat, i.e. `Y.min()` *is* the thing that really matters. (Alternatively, one could use the minimum of the posterior mean, but then we wouldn't need to touch the training data at all here). If I use `Y.max()` instead, the minimum doesn't enter the computation at all!?!?

**Comment by saitcakmak** [comment]
> PyTorch is very explicit about what affects [reproducibility](https://pytorch.org/docs/stable/notes/randomness.html) and nothing that is mentioned there seems to be a problem for this scenario. And the fact that the GP fitting yields the exact same reproducible result (from the point above) further seems to confirm this.

@esantorella looked into this quite a bit in the past. Maybe she can add more on where we've seen reproducibility issues come from in the past. IIRC, the instructions on that page weren't sufficient to ensure full reproducibility, though that may have been across different devices.

> In the end, the only thing that should enter the computation of acquisition function values is the best_f – which is identical in both cases – and the fitted multivariate normal. The latter also matches exactly for the two cases (ignoring the flip of the mean, of course), i.e. torch.equal(posterior_max.covariance_matrix, posterior_min.covariance_matrix) gives True.

Are these still identical even in the case where you've seen different acquisition values? If you can identify a random seed that reproduces the issue, it should be possible to step through the code to figure out where the values start to diverge.

At the end of the day, for evaluating an acquisition function like qLogEI, BoTorch & GPyTorch simply sequence a number of PyTorch operations. There shouldn't be any external source of randomness introduced in there. But if you can identify a bug somewhere, we'd be happy to fix it.

**Comment by esantorella** [comment]
This isn't an issue with reproducibility per se, since the same inputs produce the same outputs on repeated runs once the seed is set. The question is why negation makes a difference. The reason is that samples are being taken from two different `MultivariateNormal` distributions, one whose mean is the negated version of the other:

```python
from gpytorch.distributions import MultivariateNormal
import torch
from botorch.sampling.normal import SobolQMCNormalSampler

base_samples = torch.tensor([2.0, 4, 6]).unsqueeze(0)

loc = torch.tensor([-5, 1, 5])
covar = torch.eye(3) 

mvn_1 = MultivariateNormal(mean=loc, covariance_matrix=covar)
# [-3, 5, 11]
print(mvn_1.rsample(base_samples=base_samples))

mvn_2 = MultivariateNormal(mean=-loc, covariance_matrix=covar)
# [7, 3, 1]
print(mvn_2.rsample(base_samples=base_samples))
```

The discrepancies don't wind up too large since many samples are averaged over. Over 100 runs, I did see some large differences in the acquisition values, but only when they were very negative, so the differences in expected improvement (rather than log expected improvement) were not that large.

**Comment by AdrianSosic** [comment]
Hi @esantorella, thanks for the input, that was a very good hint indeed. So if I may summarize: you're basically saying that the MC approximation is different because the samples are not perfectly "mirrored" around 0. Instead, the base samples only undergo different shifts in mean. This makes perfect sense to me.

To verify this hypothesis, I've tested two things in my example:
1. Increasing the number of MC samples, in which case the differences should vanish. Can indeed confirm that this is the case ✅ 
2. Based on the mirroring idea, the differences should completely vanish if the base samples are truly mirrored. And this in fact turns out to be true. Below a quick'n'dirty test for this using a slightly modified version of the code, where I've completely removed the GP fitting and work with a fixed MVN posterior instead. Both outputs are now perfectly identical ✅ 

```python
import random
from typing import Optional, Union

import numpy as np
import torch
from botorch.acquisition import qLogExpectedImprovement
from botorch.acquisition.objective import LinearMCObjective, PosteriorTransform
from botorch.models.gpytorch import GPyTorchModel
from botorch.posteriors import GPyTorchPosterior, Posterior
from gpytorch.distributions import MultivariateNormal
from torch import Tensor

train_X = torch.rand(5, 2)
mean = torch.tensor([-0.2347, -0.3867, -0.1260, -0.5887, -0.0877])
covar = torch.tensor(
    [
        [0.8106, 0.7242, 0.5096, 0.0643, 0.3183],
        [0.7242, 0.6771, 0.4128, 0.0637, 0.2485],
        [0.5096, 0.4128, 0.9118, 0.0890, 0.8518],
        [0.0643, 0.0637, 0.0890, 0.0452, 0.0701],
        [0.3183, 0.2485, 0.8518, 0.0701, 0.9340],
    ]
)
best_f = torch.tensor(0.7178)


def fix_seed(seed=0):
    torch.manual_seed(seed)
    random.seed(seed)
    np.random.seed(seed)


class MirroredMultivariateNormal(MultivariateNormal):
    def __init__(self, mean, covariance_matrix, validate_args=False, mirror=False):
        self.mirror = mirror
        super().__init__(mean, covariance_matrix, validate_args)

    def rsample(self, sample_shape=..., base_samples=None):
        if self.mirror:
            base_samples = -base_samples
        return super().rsample(sample_shape, base_samples)


class DummyModel(GPyTorchModel):
    def __init__(self, posterior):
        super().__init__()
        self._posterior = posterior

    def posterior(
        self,
        X: Tensor,
        output_indices: Optional[list[int]] = None,
        observation_noise: Union[bool, Tensor] = False,
        posterior_transform: Optional[PosteriorTransform] = None,
    ) -> Posterior:
        return self._posterior


def run(minimize: bool):
    if minimize:
        obj = LinearMCObjective(weights=torch.tensor([-1.0]))
    else:
        obj = LinearMCObjective(weights=torch.tensor([1.0]))

    mvn = MirroredMultivariateNormal(
        -mean if minimize else mean, covar, mirror=minimize
    )
    posterior = GPyTorchPosterior(mvn)
    dummy_model = DummyModel(posterior)

    acqf = qLogExpectedImprovement(
        model=dummy_model,
        best_f=best_f,
        objective=obj,
    )

    fix_seed(0)
    acqf_value = acqf(train_X.unsqueeze(0))
    return posterior, acqf_value


posterior_max, acqf_max = run(False)
posterior_min, acqf_min = run(True)

assert torch.equal(posterior_max.mean, -posterior_min.mean)
assert torch.equal(posterior_max.stddev, posterior_min.stddev)

print(acqf_max.tolist())
print(acqf_min.tolist())
```

Last question from my side: do you see any simple way to achieve this perfect match without my hacks? The reason why I'm asking is because it would simplify testing.

**Comment by AdrianSosic** [comment]
PS: I think you can indeed remove the `bug` label 🙃
